miasm2/ir/translators/z3_ir.py

```diff
@@ -0,0 +1,308 @@
+import logging
+import operator
+
+import z3
+
+import miasm2.expression.expression as m2_expr
+from miasm2.ir.translators.translator import Translator
+
+log = logging.getLogger("translator_z3")
+console_handler = logging.StreamHandler()
+console_handler.setFormatter(logging.Formatter("%(levelname)-5s: %(message)s"))
+log.addHandler(console_handler)
+log.setLevel(logging.WARNING)
+
+class Z3Mem(object):
```

Comment:
In fact, it is actually the case : the implementation allows  @32[EAX] == @16[EAX] . @16[EAX + 2] for a big endian architecture.
For a little endian architecture, the equation becomes @32[EAX] == @16[EAX + 2] . @16[EAX].
This way, the memory model is coherent with memory accesses in real architectures, as far as I know.

You can verify it with this code (equiv is defined in the test section of z3_ir.py) :
```python
    mem = Z3Mem(endianness='<') # little endian                                  
    eax = z3.BitVec('EAX', 32)                                                                                                   
    assert equiv(                                                                
             # @32[EAX]                                                          
             mem.get(eax, 32),                                                   
             # @16[EAX+2] . @16[EAX]                                             
             z3.Concat(mem.get(eax+2, 16),                                       
                       mem.get(eax, 16)))
```

Any memory access is a *concatenation of bytes-sized memory access*, the direction of the concatenation depends on the endianness. To be totally precise, it is a slice of this, to handle non 8bit-aligned accesses, even though it could not make sense it is necessary to pass the automatic tests. That said, non 8bit aligned memory accesses are not tested yet, I suspect they could be wrong in some cases.

But @*16*[**EAX**] != @*16*[**AX**] even if EAX = ZeroExtend(AX), because as these addresses does not have the same size, they refer two different address spaces. This is quite a borderline case, but TranslatorZ3 should be able to translate any Expression, even if it does not necessarily make sense in a real architecture.
This can be verified with this code :

```python
    ax = z3.BitVec('AX', 16)                                                     
    assert not equiv(     
            # @16[EAX] with EAX = ZeroExtend(AX)
            mem.get(z3.ZeroExt(16, ax), 16),                                     
            mem.get(ax, 16))
```

This remarks raises two points:
- Maybe the doc is not clear enough, I will try to express it better.
- Maybe it would be usefull to allow choosing the endianness of the access in the Z3Mem.get method too, to allow getting large arrays, which you want to get with the "big endian" style. Another solution would be to add a method to get a concatenation of bytes rather than a number.

Does this solution fit the needs you raised? Do you think the API should be modified?